File: src/main/java/com/github/jactor/persistence/entity/PersistentDataEmbeddable.java
package com.github.jactor.persistence.entity;

import com.github.jactor.persistence.dto.PersistentDto;
import com.github.jactor.persistence.time.Now;
import java.time.LocalDateTime;
import javax.persistence.Embeddable;

@Embeddable
public class PersistentDataEmbeddable implements PersistentData {

  private String createdBy;
  private LocalDateTime timeOfCreation;
  private String modifiedBy;
  private LocalDateTime timeOfModification;

  PersistentDataEmbeddable() {
    createdBy = "todo";
    timeOfCreation = Now.asDateTime();
    modifiedBy = "todo";
    timeOfModification = Now.asDateTime();
  }

  PersistentDataEmbeddable(PersistentDto persistentDto) {
    createdBy = persistentDto.getCreatedBy();
    timeOfCreation = persistentDto.getTimeOfCreation();
    modifiedBy = persistentDto.getModifiedBy();
    timeOfModification = persistentDto.getTimeOfModification();
  }

  void modify() {
    timeOfModification = Now.asDateTime();
  }

  PersistentDto asPersistentDto(Long id) {
    return new PersistentDto(id, createdBy, timeOfCreation, modifiedBy, timeOfModification);
  }

  @Override
  public String getCreatedBy() {
    return createdBy;
  }

  @Override
  public LocalDateTime getTimeOfCreation() {
    return timeOfCreation;
  }

  @Override
  public String getModifiedBy() {
    return modifiedBy;
  }

  @Override
  public LocalDateTime getTimeOfModification() {
    return timeOfModification;
  }
}

File: src/main/kotlin/com/github/jactor/persistence/entity/PersistentDataEmbeddable.kt
package com.github.jactor.persistence.entity

import com.github.jactor.persistence.dto.PersistentDto
import com.github.jactor.persistence.time.Now.Companion.asDateTime
import javax.persistence.Embeddable
import java.time.LocalDateTime

@Embeddable
class PersistentDataEmbeddable : PersistentData {
    override var createdBy: String
        private set
    override var timeOfCreation: LocalDateTime
        private set
    override var modifiedBy: String
        private set
    override var timeOfModification: LocalDateTime
        private set

    protected constructor() {
        createdBy = "todo"
        timeOfCreation = asDateTime()
        modifiedBy = "todo"
        timeOfModification = asDateTime()
    }

    internal constructor(persistentDto: PersistentDto) {
        createdBy = persistentDto.createdBy
        timeOfCreation = persistentDto.timeOfCreation
        modifiedBy = persistentDto.modifiedBy
        timeOfModification = persistentDto.timeOfModification
    }

    fun modifiedBy(modifier: String) {
        modifiedBy = modifier
        timeOfModification = asDateTime()
    }

    fun asPersistentDto(id: Long?): PersistentDto {
        return PersistentDto(id, createdBy, timeOfCreation, modifiedBy, timeOfModification)
    }
}
